src/read.cpp

```diff
@@ -60,12 +61,16 @@ void read_lines_chunked_(List sourceSpec, List locale_, std::vector<std::string>
     progress,
     &locale);
 
-  CharacterVector out = r.readToVector<CharacterVector>(chunk_size);
+  CharacterVector out;
+  bool moreLines = R6method(callback, "continue")() &&
+    (out = r.readToVector<CharacterVector>(chunkSize)).size();
+
   int pos = 1;
-  while (out.size() > 0 && R6method(callback, "continue")()) {
+  while (moreLines) {
     R6method(callback, "receive")(out, pos);
     pos += out.size();
-    out = r.readToVector<CharacterVector>(chunk_size);
+    moreLines = R6method(callback, "continue")() &&
+      (out = r.readToVector<CharacterVector>(chunkSize)).size();
```

Comment:
Looks good! But isn't this also just the same as:

```cpp
while (R6method(callback, "continue")() && (out = r.readToVector<CharacterVector>(chunkSize)).size()) {
  R6method(callback, "receive")(out, pos);
  pos += out.size();
}
```

Other than that, LGTM! Nice work!